extensions/kafka/src/main/java/io/deephaven/kafka/KafkaTools.java

```diff
@@ -1312,80 +1331,183 @@ public static PartitionedTable consumeToPartitionedTable(
+         *        register} a {@link StreamConsumer} for
+         */
+        void register(
                 @NotNull TableDefinition tableDefinition,
-                @NotNull TableType tableType,
-                @NotNull Supplier<Pair<StreamToBlinkTableAdapter, ConsumerRecordToStreamPublisherAdapter>> adapterFactory,
-                @NotNull MutableObject<KafkaIngester> kafkaIngesterHolder);
+                @NotNull TopicPartition topicPartition,
+                @NotNull StreamPublisher streamPublisher);
     }
 
-    private static class TableResultFactory implements ResultFactory<Table> {
+    /**
+     * Marker interface for {@link StreamConsumer} registrar provider objects.
+     */
+    public interface StreamConsumerRegistrarProvider {
+
+        /**
+         * @param registrar The internal registrar method for {@link StreamConsumer} instances
+         * @return A StreamConsumerRegistrarProvider that registers a single consumer for all selected partitions
+         */
+        static Single single(@NotNull final SingleConsumerRegistrar registrar) {
+            return Single.of(registrar);
+        }
 
-        @Override
-        public UpdateSourceRegistrar getSourceRegistrar() {
-            return ExecutionContext.getContext().getUpdateGraph();
+        /**
+         * @param registrar The internal registrar method for {@link StreamConsumer} instances
+         * @return A StreamConsumerRegistrarProvider that registers a new consumer for each selected partition
+         */
+        static PerPartition perPartition(@NotNull final PerPartitionConsumerRegistrar registrar) {
+            return PerPartition.of(registrar);
         }
 
-        @Override
-        public Pair<Table, IntFunction<KafkaStreamConsumer>> makeResultAndConsumerFactoryPair(
-                @NotNull final TableDefinition tableDefinition,
-                @NotNull final TableType tableType,
-                @NotNull final Supplier<Pair<StreamToBlinkTableAdapter, ConsumerRecordToStreamPublisherAdapter>> adapterFactory,
-                @NotNull final MutableObject<KafkaIngester> kafkaIngesterHolder) {
-            final Pair<StreamToBlinkTableAdapter, ConsumerRecordToStreamPublisherAdapter> singleAdapterPair =
-                    adapterFactory.get();
-            final Table blinkTable = singleAdapterPair.getFirst().table();
-            final Table result = tableType.walk(new BlinkTableOperation(blinkTable));
-            final IntFunction<KafkaStreamConsumer> consumerFactory = (final int partition) -> {
-                singleAdapterPair.getFirst().setShutdownCallback(() -> kafkaIngesterHolder.getValue().shutdown());
-                return new SimpleKafkaStreamConsumer(singleAdapterPair.getSecond(), singleAdapterPair.getFirst());
-            };
-            return new Pair<>(result, consumerFactory);
+        <T, V extends Visitor<T>> T walk(V visitor);
```

Comment:
I don't think there is value in having generic `V` here:

```java
<T> T walk(Visitor<T> visitor);
```

is _good_ enough, I think?

Reply:
I was copying your pattern on `TableType`. Would you like me to change both, or am I missing some subtelty?